### PR TITLE
Use 1.51.0 in old cargos test

### PR DESCRIPTION
**Pull request by hi-rustin**
Remove TODO.
If the default toolchain is the same as "stable", just skip the `avoids_split_debuginfo_collision` test.

### CODE REVIEW

**Comment by ehuss** [comment]
Thanks!  Unfortunately it looks like these tests have atrophied over time. Would you be willing to try to get them back into running shape?  If not, that's ok, I can probably get them working again.  It might be somewhat involved to make the output conditional based on the version.

It looks like all versions from 1.51 onwards aren't working due to some output changes (and there have been various output changes over time).

I'm also wondering if it might be feasible to remove the `#[ignore]` lines to prevent that.  I can't remember if there was a really good reason for that.

**Comment by hi-rustin** [comment]
> Would you be willing to try to get them back into running shape?

I will try to fix it in the next week.

**Comment by hi-rustin** [comment]
@ehuss It's already working correctly locally.
I think the reason it was originally ignored was probably because it relied on a couple of specific versions of toolchain.
Do you think we need to run a separate GitHub actions job for it?

**Comment by hi-rustin** [comment]
> I think the reason it was originally ignored was probably because it relied on a couple of specific versions of toolchain.
> Do you think we need to run a separate GitHub actions job for it?

@ehuss Happy New Year! Could you please take a look?

**Comment by ehuss** [comment]
I'd prefer not to change CI to accommodate these tests for now.  It looks like `index_cache_rebuild` fundamentally needs 1.48 installed, so I would leave that with `#[ignore]` for now.

As for `avoids_split_debuginfo_collision`, that test needs two different toolchains. If the default toolchain is `stable`, then it won't work. I probably never tested that scenario.  Also, I think maybe the test needs to be updated due to changes in #9653.  I think one way to fix it is to check if the default toolchain is the same as "stable", just skip the test.

**Comment by hi-rustin** [comment]
Manual test results(stable): 
1. show default 
```console
➜  cargo git:(rustin-patch-test) rustup default       
stable-x86_64-apple-darwin (default)
```
2. run tests
```console
➜  cargo git:(rustin-patch-test) cargo test --test testsuite -- old_cargos --nocapture --ignored
   Compiling cargo v0.60.0 (/Users/rustin/ClionProjects/cargo)
    Finished test [unoptimized + debuginfo] target(s) in 11.31s
     Running tests/testsuite/main.rs (target/debug/deps/testsuite-7f2496f3b10b97b9)

running 3 tests
running `cargo +1.48.0 check`
test old_cargos::avoids_split_debuginfo_collision ... ok
running `/Users/rustin/ClionProjects/cargo/target/debug/cargo check -Znamespaced-features`
running `cargo +1.48.0 tree`
test old_cargos::index_cache_rebuild ... ok
1.48.0-x86_64-apple-darwin ok
1.48.0-x86_64-apple-darwin ok
1.48.0-x86_64-apple-darwin ok
1.48.0-x86_64-apple-darwin ok
1.51.0-x86_64-apple-darwin ok
1.51.0-x86_64-apple-darwin ok
1.51.0-x86_64-apple-darwin err process didn't exit successfully: `cargo +1.51.0-x86_64-apple-darwin build --verbose` (exit status: 101)
--- stderr
    Updating `/Users/rustin/ClionProjects/cargo/target/tmp/cit/t1/registry` index
error: failed to select a version for the requirement `bar = "=1.0.1"`
candidate versions found which didn't match: 1.0.2, 1.0.0
location searched: `/Users/rustin/ClionProjects/cargo/target/tmp/cit/t1/registry` index (which is replacing registry `https://github.com/rust-lang/crates.io-index`)
required by package `foo v0.1.0 (/Users/rustin/ClionProjects/cargo/target/tmp/cit/t1/foo)`
perhaps a crate was updated and forgotten to be re-vendored?

1.51.0-x86_64-apple-darwin err process didn't exit successfully: `cargo +1.51.0-x86_64-apple-darwin build --verbose` (exit status: 101)
--- stderr
    Updating `/Users/rustin/ClionProjects/cargo/target/tmp/cit/t1/registry` index
error: failed to select a version for the requirement `baz = "=1.0.1"`
candidate versions found which didn't match: 1.0.0
location searched: `/Users/rustin/ClionProjects/cargo/target/tmp/cit/t1/registry` index (which is replacing registry `https://github.com/rust-lang/crates.io-index`)
required by package `bar v1.0.2`
    ... which is depended on by `foo v0.1.0 (/Users/rustin/ClionProjects/cargo/target/tmp/cit/t1/foo)`
perhaps a crate was updated and forgotten to be re-vendored?

beta-x86_64-apple-darwin ok
beta-x86_64-apple-darwin ok
beta-x86_64-apple-darwin err process didn't exit successfully: `cargo +beta-x86_64-apple-darwin build --verbose` (exit status: 101)
--- stderr
    Updating `dummy-registry` index
error: failed to select a version for the requirement `bar = "^1.0"`
candidate versions found which didn't match: 1.0.2, 1.0.0
location searched: `dummy-registry` index (which is replacing registry `crates-io`)
required by package `foo v0.1.0 (/Users/rustin/ClionProjects/cargo/target/tmp/cit/t1/foo)`
perhaps a crate was updated and forgotten to be re-vendored?

beta-x86_64-apple-darwin err process didn't exit successfully: `cargo +beta-x86_64-apple-darwin build --verbose` (exit status: 101)
--- stderr
    Updating `dummy-registry` index
error: failed to select a version for the requirement `baz = "^1.0"`
candidate versions found which didn't match: 1.0.0
location searched: `dummy-registry` index (which is replacing registry `crates-io`)
required by package `bar v1.0.2`
    ... which satisfies dependency `bar = "^1.0"` (locked to 1.0.2) of package `foo v0.1.0 (/Users/rustin/ClionProjects/cargo/target/tmp/cit/t1/foo)`
perhaps a crate was updated and forgotten to be re-vendored?

stable-x86_64-apple-darwin ok
stable-x86_64-apple-darwin ok
stable-x86_64-apple-darwin err process didn't exit successfully: `cargo +stable-x86_64-apple-darwin build --verbose` (exit status: 101)
--- stderr
    Updating `dummy-registry` index
error: failed to select a version for the requirement `bar = "^1.0"`
candidate versions found which didn't match: 1.0.2, 1.0.0
location searched: `dummy-registry` index (which is replacing registry `crates-io`)
required by package `foo v0.1.0 (/Users/rustin/ClionProjects/cargo/target/tmp/cit/t1/foo)`
perhaps a crate was updated and forgotten to be re-vendored?

stable-x86_64-apple-darwin err process didn't exit successfully: `cargo +stable-x86_64-apple-darwin build --verbose` (exit status: 101)
--- stderr
    Updating `dummy-registry` index
error: failed to select a version for the requirement `baz = "^1.0"`
candidate versions found which didn't match: 1.0.0
location searched: `dummy-registry` index (which is replacing registry `crates-io`)
required by package `bar v1.0.2`
    ... which satisfies dependency `bar = "^1.0"` (locked to 1.0.2) of package `foo v0.1.0 (/Users/rustin/ClionProjects/cargo/target/tmp/cit/t1/foo)`
perhaps a crate was updated and forgotten to be re-vendored?

this ok
this ok
this err process didn't exit successfully: `/Users/rustin/ClionProjects/cargo/target/debug/cargo build --verbose` (exit status: 101)
--- stderr
    Updating `dummy-registry` index
error: failed to select a version for the requirement `bar = "^1.0"`
candidate versions found which didn't match: 1.0.2, 1.0.0
location searched: `dummy-registry` index (which is replacing registry `crates-io`)
required by package `foo v0.1.0 (/Users/rustin/ClionProjects/cargo/target/tmp/cit/t1/foo)`
perhaps a crate was updated and forgotten to be re-vendored?

this err process didn't exit successfully: `/Users/rustin/ClionProjects/cargo/target/debug/cargo build --verbose` (exit status: 101)
--- stderr
    Updating `dummy-registry` index
error: failed to select a version for the requirement `baz = "^1.0"`
candidate versions found which didn't match: 1.0.0
location searched: `dummy-registry` index (which is replacing registry `crates-io`)
required by package `bar v1.0.2`
    ... which satisfies dependency `bar = "^1.0"` (locked to 1.0.2) of package `foo v0.1.0 (/Users/rustin/ClionProjects/cargo/target/tmp/cit/t1/foo)`
perhaps a crate was updated and forgotten to be re-vendored?

nightly-x86_64-apple-darwin ok
nightly-x86_64-apple-darwin ok
nightly-x86_64-apple-darwin err process didn't exit successfully: `cargo +nightly-x86_64-apple-darwin build --verbose` (exit status: 101)
--- stderr
    Updating `dummy-registry` index
error: failed to select a version for the requirement `bar = "^1.0"`
candidate versions found which didn't match: 1.0.2, 1.0.0
location searched: `dummy-registry` index (which is replacing registry `crates-io`)
required by package `foo v0.1.0 (/Users/rustin/ClionProjects/cargo/target/tmp/cit/t1/foo)`
perhaps a crate was updated and forgotten to be re-vendored?

nightly-x86_64-apple-darwin err process didn't exit successfully: `cargo +nightly-x86_64-apple-darwin build --verbose` (exit status: 101)
--- stderr
    Updating `dummy-registry` index
error: failed to select a version for the requirement `baz = "^1.0"`
candidate versions found which didn't match: 1.0.0
location searched: `dummy-registry` index (which is replacing registry `crates-io`)
required by package `bar v1.0.2`
    ... which satisfies dependency `bar = "^1.0"` (locked to 1.0.2) of package `foo v0.1.0 (/Users/rustin/ClionProjects/cargo/target/tmp/cit/t1/foo)`
perhaps a crate was updated and forgotten to be re-vendored?

test old_cargos::new_features ... ok

test result: ok. 3 passed; 0 failed; 0 ignored; 0 measured; 2286 filtered out; finished in 26.96s
```

**Comment by hi-rustin** [comment]
Manual test results(beta): 
1. show default 
```console
➜  cargo git:(rustin-patch-test) rustup default                                                 
beta-x86_64-apple-darwin (default)
```
2. run tests
```console
cargo git:(rustin-patch-test) cargo test --test testsuite -- old_cargos --nocapture --ignored
   Compiling cargo v0.60.0 (/Users/rustin/ClionProjects/cargo)
    Finished test [unoptimized + debuginfo] target(s) in 14.52s
     Running tests/testsuite/main.rs (target/debug/deps/testsuite-56879c9ac5a47962)

running 3 tests
running `cargo +1.48.0 check`
running `cargo +stable build`
running `/Users/rustin/ClionProjects/cargo/target/debug/cargo check -Znamespaced-features`
running `cargo +1.48.0 tree`
1.48.0-x86_64-apple-darwin ok
test old_cargos::index_cache_rebuild ... ok
running `/Users/rustin/ClionProjects/cargo/target/debug/cargo build`
running `cargo +stable build`
1.48.0-x86_64-apple-darwin ok
1.48.0-x86_64-apple-darwin ok
1.48.0-x86_64-apple-darwin ok
test old_cargos::avoids_split_debuginfo_collision ... ok
1.51.0-x86_64-apple-darwin ok
1.51.0-x86_64-apple-darwin ok
1.51.0-x86_64-apple-darwin err process didn't exit successfully: `cargo +1.51.0-x86_64-apple-darwin build --verbose` (exit status: 101)
--- stderr
    Updating `/Users/rustin/ClionProjects/cargo/target/tmp/cit/t0/registry` index
error: failed to select a version for the requirement `bar = "=1.0.1"`
candidate versions found which didn't match: 1.0.2, 1.0.0
location searched: `/Users/rustin/ClionProjects/cargo/target/tmp/cit/t0/registry` index (which is replacing registry `https://github.com/rust-lang/crates.io-index`)
required by package `foo v0.1.0 (/Users/rustin/ClionProjects/cargo/target/tmp/cit/t0/foo)`
perhaps a crate was updated and forgotten to be re-vendored?

1.51.0-x86_64-apple-darwin err process didn't exit successfully: `cargo +1.51.0-x86_64-apple-darwin build --verbose` (exit status: 101)
--- stderr
    Updating `/Users/rustin/ClionProjects/cargo/target/tmp/cit/t0/registry` index
error: failed to select a version for the requirement `baz = "=1.0.1"`
candidate versions found which didn't match: 1.0.0
location searched: `/Users/rustin/ClionProjects/cargo/target/tmp/cit/t0/registry` index (which is replacing registry `https://github.com/rust-lang/crates.io-index`)
required by package `bar v1.0.2`
    ... which is depended on by `foo v0.1.0 (/Users/rustin/ClionProjects/cargo/target/tmp/cit/t0/foo)`
perhaps a crate was updated and forgotten to be re-vendored?

beta-x86_64-apple-darwin ok
beta-x86_64-apple-darwin ok
beta-x86_64-apple-darwin err process didn't exit successfully: `cargo +beta-x86_64-apple-darwin build --verbose` (exit status: 101)
--- stderr
    Updating `dummy-registry` index
error: failed to select a version for the requirement `bar = "^1.0"`
candidate versions found which didn't match: 1.0.2, 1.0.0
location searched: `dummy-registry` index (which is replacing registry `crates-io`)
required by package `foo v0.1.0 (/Users/rustin/ClionProjects/cargo/target/tmp/cit/t0/foo)`
perhaps a crate was updated and forgotten to be re-vendored?

beta-x86_64-apple-darwin err process didn't exit successfully: `cargo +beta-x86_64-apple-darwin build --verbose` (exit status: 101)
--- stderr
    Updating `dummy-registry` index
error: failed to select a version for the requirement `baz = "^1.0"`
candidate versions found which didn't match: 1.0.0
location searched: `dummy-registry` index (which is replacing registry `crates-io`)
required by package `bar v1.0.2`
    ... which satisfies dependency `bar = "^1.0"` (locked to 1.0.2) of package `foo v0.1.0 (/Users/rustin/ClionProjects/cargo/target/tmp/cit/t0/foo)`
perhaps a crate was updated and forgotten to be re-vendored?

this ok
this ok
this err process didn't exit successfully: `/Users/rustin/ClionProjects/cargo/target/debug/cargo build --verbose` (exit status: 101)
--- stderr
    Updating `dummy-registry` index
error: failed to select a version for the requirement `bar = "^1.0"`
candidate versions found which didn't match: 1.0.2, 1.0.0
location searched: `dummy-registry` index (which is replacing registry `crates-io`)
required by package `foo v0.1.0 (/Users/rustin/ClionProjects/cargo/target/tmp/cit/t0/foo)`
perhaps a crate was updated and forgotten to be re-vendored?

this err process didn't exit successfully: `/Users/rustin/ClionProjects/cargo/target/debug/cargo build --verbose` (exit status: 101)
--- stderr
    Updating `dummy-registry` index
error: failed to select a version for the requirement `baz = "^1.0"`
candidate versions found which didn't match: 1.0.0
location searched: `dummy-registry` index (which is replacing registry `crates-io`)
required by package `bar v1.0.2`
    ... which satisfies dependency `bar = "^1.0"` (locked to 1.0.2) of package `foo v0.1.0 (/Users/rustin/ClionProjects/cargo/target/tmp/cit/t0/foo)`
perhaps a crate was updated and forgotten to be re-vendored?

stable-x86_64-apple-darwin ok
stable-x86_64-apple-darwin ok
stable-x86_64-apple-darwin err process didn't exit successfully: `cargo +stable-x86_64-apple-darwin build --verbose` (exit status: 101)
--- stderr
    Updating `dummy-registry` index
error: failed to select a version for the requirement `bar = "^1.0"`
candidate versions found which didn't match: 1.0.2, 1.0.0
location searched: `dummy-registry` index (which is replacing registry `crates-io`)
required by package `foo v0.1.0 (/Users/rustin/ClionProjects/cargo/target/tmp/cit/t0/foo)`
perhaps a crate was updated and forgotten to be re-vendored?

stable-x86_64-apple-darwin err process didn't exit successfully: `cargo +stable-x86_64-apple-darwin build --verbose` (exit status: 101)
--- stderr
    Updating `dummy-registry` index
error: failed to select a version for the requirement `baz = "^1.0"`
candidate versions found which didn't match: 1.0.0
location searched: `dummy-registry` index (which is replacing registry `crates-io`)
required by package `bar v1.0.2`
    ... which satisfies dependency `bar = "^1.0"` (locked to 1.0.2) of package `foo v0.1.0 (/Users/rustin/ClionProjects/cargo/target/tmp/cit/t0/foo)`
perhaps a crate was updated and forgotten to be re-vendored?

nightly-x86_64-apple-darwin ok
nightly-x86_64-apple-darwin ok
nightly-x86_64-apple-darwin err process didn't exit successfully: `cargo +nightly-x86_64-apple-darwin build --verbose` (exit status: 101)
--- stderr
    Updating `dummy-registry` index
error: failed to select a version for the requirement `bar = "^1.0"`
candidate versions found which didn't match: 1.0.2, 1.0.0
location searched: `dummy-registry` index (which is replacing registry `crates-io`)
required by package `foo v0.1.0 (/Users/rustin/ClionProjects/cargo/target/tmp/cit/t0/foo)`
perhaps a crate was updated and forgotten to be re-vendored?

nightly-x86_64-apple-darwin err process didn't exit successfully: `cargo +nightly-x86_64-apple-darwin build --verbose` (exit status: 101)
--- stderr
    Updating `dummy-registry` index
error: failed to select a version for the requirement `baz = "^1.0"`
candidate versions found which didn't match: 1.0.0
location searched: `dummy-registry` index (which is replacing registry `crates-io`)
required by package `bar v1.0.2`
    ... which satisfies dependency `bar = "^1.0"` (locked to 1.0.2) of package `foo v0.1.0 (/Users/rustin/ClionProjects/cargo/target/tmp/cit/t0/foo)`
perhaps a crate was updated and forgotten to be re-vendored?

test old_cargos::new_features ... ok

test result: ok. 3 passed; 0 failed; 0 ignored; 0 measured; 2286 filtered out; finished in 27.76s
```

**Comment by hi-rustin** [comment]
@ehuss 
Could you please take a look? Thanks! 💚 💙 💜 💛 ❤️

**Comment by hi-rustin** [comment]
nightly also work!

File: tests/testsuite/old_cargos.rs [hunk 1]
```diff
@@ -92,6 +92,25 @@ fn collect_all_toolchains() -> Vec<(Version, String)> {
     toolchains
 }
 
+/// Returns whether the default toolchain is the stable version.
+fn default_toolchain_is_stable() -> bool {
```
Comment (ehuss) [inline tests/testsuite/old_cargos.rs:96]: Could this perhaps be simplified a little?  Maybe something like this:

```rust
    let default = tc_process("rustc", "this").arg("-V").exec_with_output();
    let stable = tc_process("rustc", "stable").arg("-V").exec_with_output();
    match (default, stable) {
        (Ok(d), Ok(s)) => d.stdout == s.stdout,
        _ => false,
    }
```

File: tests/testsuite/old_cargos.rs [hunk 2]
```diff
@@ -591,6 +611,11 @@ foo v0.1.0 [..]
 #[cargo_test]
 #[ignore]
 fn avoids_split_debuginfo_collision() {
+    // Test needs two different toolchains.
+    // If the default toolchain is stable, then it won't work.
```
Comment (ehuss) [inline tests/testsuite/old_cargos.rs:615]: There's a non-breaking space here.

```suggestion
    // If the default toolchain is stable, then it won't work.
```

**Review by ehuss — COMMENTED** [review]
Sorry about the delay.

It looks like `new_features` has fallen behind again.  Starting with 1.60, namespaced-features has been stabilized. That means the validity checks will need to have three different checks:

* < 1.51
* >= 1.51 && < 1.60
* >= 1.60

Can you add the new 1.60 behavior?

**Comment by ehuss** [comment]
Thanks!

@bors r+

**Pull request merged**